x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -727,16 +814,313 @@ public void runAsDenied(String requestId, Authentication authentication, RestReq
+        }
+
+        private void withRoleDescriptor(XContentBuilder builder, RoleDescriptor roleDescriptor) throws IOException {
+            builder.startObject()
+                .array(RoleDescriptor.Fields.CLUSTER.getPreferredName(), roleDescriptor.getClusterPrivileges());
+            if (roleDescriptor.getConditionalClusterPrivileges() != null && roleDescriptor.getConditionalClusterPrivileges().length > 0) {
+                // This fails if this list contains multiple instances of the {@code ManageApplicationPrivileges}
+                // Again, only the transport client can produce this, and this only introduces a different failure mode and
+                // not a new one (i.e. without auditing it would fail differently, but it would still fail)
+                builder.field(RoleDescriptor.Fields.GLOBAL.getPreferredName());
+                ConfigurableClusterPrivileges.toXContent(builder, ToXContent.EMPTY_PARAMS,
+                        Arrays.asList(roleDescriptor.getConditionalClusterPrivileges()));
+            }
+            builder.startArray(RoleDescriptor.Fields.INDICES.getPreferredName());
+            for (RoleDescriptor.IndicesPrivileges indicesPrivileges : roleDescriptor.getIndicesPrivileges()) {
+                withIndicesPrivileges(builder, indicesPrivileges);
+            }
+            builder.endArray();
+            // the toXContent method of the {@code RoleDescriptor.ApplicationResourcePrivileges) does a good job
+            builder.array(RoleDescriptor.Fields.APPLICATIONS.getPreferredName(), (Object[]) roleDescriptor.getApplicationPrivileges());
+            builder.array(RoleDescriptor.Fields.RUN_AS.getPreferredName(), roleDescriptor.getRunAs());
+            if (roleDescriptor.getMetadata() != null && false == roleDescriptor.getMetadata().isEmpty()) {
+                // JSON building for the metadata might fail when encountering unknown class types.
+                // This is NOT a problem because such metadata (eg containing GeoPoint) will most probably
+                // cause troubles in downstream code (eg storing the metadata), so this simply introduces a new failure mode.
+                // Also the malevolent metadata can only be produced by the transport client.
+                builder.field(RoleDescriptor.Fields.METADATA.getPreferredName(), roleDescriptor.getMetadata());
+            }
+            builder.endObject();
+        }
+
+        private void withIndicesPrivileges(XContentBuilder builder, RoleDescriptor.IndicesPrivileges indicesPrivileges) throws IOException {
+            builder.startObject();
+            builder.array("names", indicesPrivileges.getIndices());
+            builder.array("privileges", indicesPrivileges.getPrivileges());
+            if (indicesPrivileges.isUsingFieldLevelSecurity()) {
+                builder.startObject(RoleDescriptor.Fields.FIELD_PERMISSIONS.getPreferredName());
+                if (indicesPrivileges.hasGrantedFields()) {
```

Comment:
Yes, I agree.
I will make it so that `grant` is always printed if `isUsingFieldLevelSecurity` is `true`.
The other checks along the way should prevent inapropriate `null` or `[]` values, but there's no biggie if they don't.

Reply:
I have pushed https://github.com/elastic/elasticsearch/pull/62916/commits/786518ba128664a60731e7cd98b6b50293dcf7ee